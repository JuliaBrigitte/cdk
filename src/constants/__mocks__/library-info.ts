const mockAwsCdkVersion = "0.0.1";

export const LibraryInfo = {
  VERSION: "1.0.0",
  AWS_CDK_VERSION: mockAwsCdkVersion,
  AWS_CDK_VERSIONS: {
    "aws-cdk": mockAwsCdkVersion,
    "@aws-cdk/core": mockAwsCdkVersion,
  },
};
