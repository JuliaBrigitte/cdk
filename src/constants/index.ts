export * from "./regex-pattern";
export * from "./stage";
