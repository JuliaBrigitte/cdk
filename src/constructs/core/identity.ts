export interface Identity {
  stack: string;
  stage: string;
  app: string;
}
