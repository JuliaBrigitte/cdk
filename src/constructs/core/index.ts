export * from "./identity";
export * from "./mappings";
export * from "./migrating";
export * from "./parameters";
export * from "./ssm";
export * from "./stack";
