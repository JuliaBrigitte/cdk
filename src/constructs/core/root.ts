import type { StageSynthesisOptions } from "aws-cdk-lib";
import { App } from "aws-cdk-lib";
import type { CloudAssembly } from "aws-cdk-lib/cx-api";
import { RiffRaffYamlFile } from "../../utils/riff-raff-yaml-file";

// Deliberately not called `GuApp` as `app` is overloaded (as is `stack`, and `stage` ðŸ˜…)
export class GuRoot extends App {
  override synth(options?: StageSynthesisOptions): CloudAssembly {
    new RiffRaffYamlFile(this).synth();
    return super.synth(options);
  }
}
