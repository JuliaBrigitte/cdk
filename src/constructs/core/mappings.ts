import { CfnMapping } from "@aws-cdk/core";
import type { GuStack } from "./stack";

export interface GuStageDependentValue {
  variableName: string;
  codeValue: unknown;
  prodValue: unknown;
}

export class GuStageMapping extends CfnMapping {
  constructor(scope: GuStack, id: string = "stage-mapping") {
    super(scope, id);
  }
}
