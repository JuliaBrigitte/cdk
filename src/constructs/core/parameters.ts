import type { CfnParameterProps, Construct } from "@aws-cdk/core";
import { CfnParameter } from "@aws-cdk/core";
import { Stage, Stages } from "../../constants";

export type GuParameterProps = CfnParameterProps;

export type GuNoTypeParameterProps = Omit<GuParameterProps, "type">;

export class GuParameter extends CfnParameter {
  static defaultProps: GuParameterProps = {};

  constructor(scope: Construct, id: string, props: GuParameterProps) {
    super(scope, id, { ...GuParameter.defaultProps, ...props });
  }
}

export class GuStringParameter extends GuParameter {
  constructor(scope: Construct, id: string, props: GuNoTypeParameterProps) {
    super(scope, id, { ...props, type: "String" });
  }
}

export class GuStageParameter extends GuParameter {
  constructor(scope: Construct) {
    super(scope, "Stage", {
      type: "String",
      description: "Stage name",
      allowedValues: Stages,
      default: Stage.CODE,
    });
  }
}

export class GuStackParameter extends GuParameter {
  constructor(scope: Construct) {
    super(scope, "Stack", {
      type: "String",
      description: "Name of this stack",
      default: "deploy",
    });
  }
}

// TODO: Is there a way of removing default props if they weren't implemented before?
//       Should that be allowed even if it is possible?
export class GuInstanceTypeParameter extends GuParameter {
  static defaultProps: GuParameterProps = {
    type: "String",
    description: "EC2 Instance Type",
    default: "t3.small",
  };

  constructor(scope: Construct, id: string = "InstanceType", props: GuParameterProps = {}) {
    super(scope, id, { ...GuInstanceTypeParameter.defaultProps, ...props });
  }
}

export class GuSSMParameter extends GuParameter {
  static defaultProps: GuParameterProps = {
    noEcho: true,
  };

  constructor(scope: Construct, id: string, props: GuNoTypeParameterProps) {
    super(scope, id, {
      ...GuSSMParameter.defaultProps,
      ...props,
      type: "AWS::SSM::Parameter::Value<String>",
    });
  }
}

export class GuSubnetListParameter extends GuParameter {
  static defaultProps: GuParameterProps = {};

  constructor(scope: Construct, id: string, props: GuNoTypeParameterProps) {
    super(scope, "Subnets", {
      ...GuSubnetListParameter.defaultProps,
      ...props,
      type: "List<AWS::EC2::Subnet::Id>",
    });
  }
}

export class GuVpcParameter extends GuParameter {
  static defaultProps: GuParameterProps = {};

  constructor(scope: Construct, id: string, props: GuNoTypeParameterProps) {
    super(scope, id, {
      ...GuVpcParameter.defaultProps,
      ...props,
      type: "AWS::EC2::VPC::Id",
    });
  }
}

export class GuAmiParameter extends GuParameter {
  static defaultProps: GuParameterProps = {};

  constructor(scope: Construct, id: string, props: GuNoTypeParameterProps) {
    super(scope, id, {
      ...GuAmiParameter.defaultProps,
      ...props,
      type: "AWS::EC2::Image::Id",
    });
  }
}
