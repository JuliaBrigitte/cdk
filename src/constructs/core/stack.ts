import type { App, StackProps } from "@aws-cdk/core";
import { Stack, Tags } from "@aws-cdk/core";
import { GuStackParameter, GuStageParameter } from "./parameters";

export interface GuStackProps extends StackProps {
  // This limits GuStack to supporting a single app.
  // In the future, support for stacks with multiple apps may be required
  app: string;
  migratedFromCloudFormation?: boolean;
}

export class GuStack extends Stack {
  private readonly _stage: GuStageParameter;
  private readonly _stack: GuStackParameter;
  private readonly _app: string;

  get stage(): string {
    return this._stage.valueAsString;
  }

  get stack(): string {
    return this._stack.valueAsString;
  }

  get app(): string {
    return this._app;
  }

  migratedFromCloudFormation: boolean;

  protected addTag(key: string, value: string, applyToLaunchedInstances: boolean = true): void {
    Tags.of(this).add(key, value, { applyToLaunchedInstances });
  }

  // eslint-disable-next-line custom-rules/valid-constructors -- GuStack is the exception as it must take an App
  constructor(app: App, id: string, props: GuStackProps) {
    super(app, id, props);

    this.migratedFromCloudFormation = !!props.migratedFromCloudFormation;

    this._stage = new GuStageParameter(this);
    this._stack = new GuStackParameter(this);
    this._app = props.app;

    this.addTag("Stack", this.stack);
    this.addTag("Stage", this.stage);
    this.addTag("App", props.app);
  }
}
