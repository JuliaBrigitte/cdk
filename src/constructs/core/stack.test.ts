import "@aws-cdk/assert/jest";

import { SynthUtils } from "@aws-cdk/assert";
import { Role, ServicePrincipal } from "@aws-cdk/aws-iam";
import { App } from "@aws-cdk/core";
import { alphabeticalTags, simpleGuStackForTesting } from "../../../test/utils";
import type { SynthedStack } from "../../../test/utils";
import { Stage, Stages } from "../../constants";
import { TrackingTag } from "../../constants/library-info";
import { GuStack } from "./stack";

describe("The GuStack construct", () => {
  it("should be able to take stack and stage values as props", function () {
    const stack = simpleGuStackForTesting({ stack: "some-stack", stage: "some-stage" });
    expect(stack.stack).toEqual("some-stack");
    expect(stack.stage).toEqual("some-stage");

    const json = SynthUtils.toCloudFormation(stack) as SynthedStack;

    expect(json.Parameters).toBeUndefined();
  });

  it("can accept only one of stack or stage", function () {
    const stack = simpleGuStackForTesting({ stage: "some-stage" });
    expect(stack.stage).toEqual("some-stage");
  });

  it("should have stack and stage parameters", () => {
    const stack = simpleGuStackForTesting();

    const json = SynthUtils.toCloudFormation(stack) as SynthedStack;

    expect(json.Parameters).toEqual({
      Stack: {
        Type: "String",
        Description: "Name of this stack",
        Default: "deploy",
      },
      Stage: {
        Type: "String",
        Description: "Stage name",
        AllowedValues: Stages,
        Default: Stage.CODE,
      },
    });
  });

  it("should apply the stack, stage and app tags to resources added to it", () => {
    const stack = new GuStack(new App(), "Test", { app: "MyApp" });

    new Role(stack, "MyRole", {
      assumedBy: new ServicePrincipal("ec2.amazonaws.com"),
    });

    expect(stack).toHaveResource("AWS::IAM::Role", {
      Tags: alphabeticalTags([
        {
          Key: "App",
          Value: "MyApp",
        },
        {
          Key: "Stack",
          Value: {
            Ref: "Stack",
          },
        },
        {
          Key: "Stage",
          Value: {
            Ref: "Stage",
          },
        },
        TrackingTag,
      ]),
    });
  });

  it("should return the correct app value when app is set", () => {
    const stack = new GuStack(new App(), "Test", { app: "MyApp" });

    expect(stack.app).toBe("MyApp");
  });
});
