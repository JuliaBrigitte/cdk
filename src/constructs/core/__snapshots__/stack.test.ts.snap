// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GuStack should apply the stack and stage tags to resources added to it 1`] = `
Object {
  "Parameters": Object {
    "Stack": Object {
      "Default": "deploy",
      "Description": "Name of this stack",
      "Type": "String",
    },
    "Stage": Object {
      "AllowedValues": Array [
        "CODE",
        "PROD",
      ],
      "Default": "CODE",
      "Description": "Stage name",
      "Type": "String",
    },
  },
  "Resources": Object {
    "MyRoleF48FFE04": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "ec2.",
                      Object {
                        "Ref": "AWS::URLSuffix",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Tags": Array [
          Object {
            "Key": "Stack",
            "Value": Object {
              "Ref": "Stack",
            },
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
